ofdb-entities/src/email.rs

```diff
@@ -1,62 +1,63 @@
-use std::{borrow::Borrow, fmt, ops::Deref, str::FromStr};
+use std::{fmt, str::FromStr};
+use thiserror::Error;
 
-// TODO: rename to EmailAddress
 #[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
-pub struct Email(String);
-
-impl AsRef<String> for Email {
-    fn as_ref(&self) -> &String {
-        &self.0
-    }
+pub struct EmailAddress {
+    address: String,
+    display_name: Option<String>,
 }
 
-impl AsRef<str> for Email {
-    fn as_ref(&self) -> &str {
-        self.0.as_str()
+impl EmailAddress {
+    pub const fn new_unchecked(address: String) -> Self {
+        Self {
+            address,
+            display_name: None,
+        }
     }
-}
-
-impl From<Email> for String {
-    fn from(from: Email) -> Self {
-        from.0
+    pub fn into_string(self) -> String {
+        self.address
     }
-}
-
-impl From<String> for Email {
-    fn from(from: String) -> Self {
-        Self(from)
+    pub fn as_str(&self) -> &str {
+        self.address.as_str()
     }
-}
-
-impl From<&str> for Email {
-    fn from(from: &str) -> Self {
-        from.to_owned().into()
+    pub fn len(&self) -> usize {
+        self.address.len()
     }
-}
-
-impl FromStr for Email {
-    type Err = ();
-    fn from_str(s: &str) -> Result<Email, Self::Err> {
-        Ok(s.into())
+    pub fn is_empty(&self) -> bool {
+        self.address.is_empty()
     }
 }
 
-impl Borrow<str> for Email {
-    fn borrow(&self) -> &str {
-        self.as_ref()
+#[derive(Debug, Error)]
+#[error("Invalid E-Mail address")]
+pub struct EmailAddressParseError;
+
+impl FromStr for EmailAddress {
+    type Err = EmailAddressParseError;
+    fn from_str(s: &str) -> Result<EmailAddress, Self::Err> {
+        let info = mailparse::addrparse(s)
+            .ok()
+            .and_then(|list| list.extract_single_info())
+            .ok_or(EmailAddressParseError)?;
+        Ok(Self {
+            address: info.addr,
+            display_name: info.display_name,
+        })
     }
 }
 
-impl Deref for Email {
-    type Target = String;
-
-    fn deref(&self) -> &String {
-        self.as_ref()
+impl fmt::Display for EmailAddress {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(name) = &self.display_name {
+            write!(f, r#""{}" <{}>"#, name.replace('"', r#"\""#), self.address)
```

Comment:
Would be easier to read with destructuring and named arguments even with temporary transformations like `display_name = display_name.replace('"', r#"\""#)`. Keep the property names as is.

Reply:
This is copy&paste from [mailparse](https://docs.rs/mailparse/0.13.8/src/mailparse/addrparse.rs.html#31-39) and I agree :wink: